src/library/scala/collection/Iterable.scala

```diff
@@ -549,21 +549,22 @@ trait IterableOps[+A, +CC[_], +C] extends Any with IterableOnce[A] with Iterable
     *
     */
   def groupBy[K](f: A => K): immutable.Map[K, C] = {
-    val m = mutable.Map.empty[K, Builder[A, C]]
+    val m = mutable.HashMap.empty[K, Any /*C | Builder[A, C] */]
     val it = iterator
-    while (it.hasNext) {
-      val elem = it.next()
-      val key = f(elem)
-      val bldr = m.getOrElseUpdate(key, newSpecificBuilder)
-      bldr += elem
-    }
-    var result = immutable.HashMap.empty[K, C]
-    val mapIt = m.iterator
-    while (mapIt.hasNext) {
-      val (k, v) = mapIt.next()
-      result = result.updated(k, v.result())
+    if (it.isEmpty) immutable.Map.empty
+    else {
+      while (it.hasNext) {
+        val elem = it.next()
+        val key  = f(elem)
+        val bldr = m.getOrElseUpdate(key, newSpecificBuilder).asInstanceOf[Builder[A, C]]
+        bldr += elem
+      }
+      m.mapValuesInPlace((_, v) => v.asInstanceOf[Builder[A, C]].result())
+      val m1 = immutable.HashMap.newBuilder[K, C]
```

Comment:
do we care if the result is a HashMap. Currently the result could be a Map1-4 as well